src/lem/circuit.rs

```diff
@@ -1007,26 +1171,35 @@ impl Func {
                     .with_context(|| " couldn't constrain `enforce_selector_with_premise`")
                 }
                 Ctrl::MatchVal(match_var, cases, def) => {
-                    let match_lit = bound_allocations.get(match_var)?.hash().clone();
+                    let match_lit = bound_allocations.get(match_var)?.clone();
                     let mut selector = Vec::with_capacity(cases.len() + 2);
                     let mut branch_slots = Vec::with_capacity(cases.len());
                     for (i, (lit, block)) in cases.iter().enumerate() {
-                        let lit_ptr = lit.to_ptr(g.store);
+                        let lit_ptr = lit.to_ptr_cached(g.store);
+                        let lit_tag = g.store.hash_ptr(&lit_ptr)?.tag.to_field::<F>();
                         let lit_hash = g.store.hash_ptr(&lit_ptr)?.hash;
-                        let is_eq = not_dummy.get_value().and_then(|not_dummy| {
+
+                        let has_match_bool = not_dummy.get_value().and_then(|not_dummy| {
                             match_lit
                                 .get_value()
-                                .map(|val| not_dummy && val == lit_hash)
+                                .map(|(tag, hash)| not_dummy && tag == lit_tag && hash == lit_hash)
                         });
-
                         let has_match = Boolean::Is(AllocatedBit::alloc(
                             &mut cs.namespace(|| format!("{i}.allocated_bit")),
-                            is_eq,
+                            has_match_bool,
```

Comment:
This appears to be an unconstrained evaluation.

Reply:
So, this comment and the above one, about `dummy` being added to the selector, has to do with the match optimization, which was merged.

Here's how matches work:
We allocate one boolean for each match arm. Then we add, to each match arm, certain constrains. For instance, in
```
match x.tag {
  Expr::Cons => ...,
  Expr::Nil => ...
  _ => ...
}
```
the first arm has a constrain `b1 -> x.tag = Expr::Cons` (an `implies_equal`), the second has `b2 -> x.tag = Expr::Nil` and the third one, which is the default, has a constrain `b3 -> (x.tag != Expr::Cons & x.tag != Expr::Nil)`. I.e., we are, at the point of allocating the booleans, in the freedom to "choose" which one to take, except that we must show right afterwards that if the boolean is true, then the variable matches the pattern.

Now, this leaves out one important thing: the `not_dummy`. But here's how it works: if `not_dummy` is FALSE (`dummy` is TRUE), then none of the branches should be chosen. That's why I added `dummy` to the popcount, so that if it is true, then none of the branches should be taken. Notice that if it wasn't for the need for accounting for dummies, a popcount is not necessary, since I made sure that the branches are always mutually exclusive